src/event_loop_integration.rs

```diff
@@ -0,0 +1,78 @@
+//! # Integrating x11rb with an Event Loop
+//!
+//! To integrate x11rb with an event loop,
+//! [`std::os::unix::io::AsRawFd`](https://doc.rust-lang.org/std/os/unix/io/trait.AsRawFd.html) is
+//! implemented by [`RustConnection`](../rust_connection/struct.RustConnection.html)'s
+//! [`DefaultStream`](../rust_connection/struct.DefaultStream.html#impl-AsRawFd) and
+//! [`XCBConnection`](../xcb_ffi/struct.XCBConnection.html#impl-AsRawFd). This allows to integrate
+//! with an event loop that also handles timeouts or network I/O. See
+//! [`xclock_utc`](https://github.com/psychon/x11rb/blob/master/examples/xclock_utc.rs) for an
+//! example.
+//!
+//! The general form of such an integration could be as follows:
+//! ```no_run
+//! #[cfg(unix)]
+//! use std::os::unix::io::{AsRawFd, RawFd};
+//! #[cfg(windows)]
+//! use std::os::windows::io::{AsRawSocket, RawSocket};
+//! use x11rb::connection::Connection;
+//! use x11rb::rust_connection::RustConnection;
+//! use x11rb::errors::ConnectionError;
+//!
+//! fn main_loop(conn: &RustConnection) -> Result<(), ConnectionError> {
+//!     #[cfg(unix)]
+//!     let raw_handle = conn.stream().as_raw_fd();
+//!     #[cfg(windows)]
+//!     let raw_handle = conn.stream().as_raw_socket();
+//!     loop {
+//!         while let Some(event) = conn.poll_for_event()? {
+//!             handle_event(event);
+//!         }
+//!
+//!         poll_for_readable(raw_handle);
+//!
+//!         // Do other work here.
+//!     }
+//! }
+//! # fn handle_event<T>(event: T) {}
+//! # fn poll_for_readable<T>(event: T) {}
+//! ```
+//! The function `poll_for_readable` could wait for any number of I/O streams (besides the one from
+//! x11rb) to become readable. It can also implement timeouts, as seen in the
+//! [`xclock_utc` example](https://github.com/psychon/x11rb/blob/master/examples/xclock_utc.rs).
+//!
+//!
+//! ## Threads and Races
+//!
+//! Both [`RustConnection`](../rust_connection/struct.RustConnection.html) and
+//! [`XCBConnection`](../xcb_ffi/struct.XCBConnection.html) are `Sync+Send`. However, it is still
+//! possible to see races in the presence of threads and an event loop.
+//!
+//! The underlying problem is that the following two points are not equivalent:
+//!
+//! 1. A new event is available and can be returned from `conn.poll_for_event()`.
+//! 2. The underlying I/O stream is readable.
```

Comment:
Uhm... no. "2 => 1" being wrong is the whole point of what we are talking about. Or did I get the order wrong here...?

"1 => 2" is also wrong, because the packet that is available could be something else than an event, i.e. an X11 error or a reply. (When you `drop` a cookie for a request with a reply, the reply is still generated, but the `Drop` impl arranges for x11rb to just ignore the reply when it comes in, but it still comes in.)
However, this case only means that the following is possible: "socket is readable, so let's `poll_for_event()`... ohh, I got `None`". That should not really cause any issues.

Reply:
Either you did get the order wrong or my coffee wasn't strong enough this morning. But you answered my (intended) question anyway, thanks!

Reply:
Your coffee was strong enough. The problem is that I do not drink coffee. Sorry for getting the order wrong. :disappointed:

Reply:
:coffee: